src/xenguestlib/Wmi.cs

```
@@ -634,19 +628,79 @@ public class WmiSession : IDisposable
         private ManagementObject session = null;
         private WmiBase wmibase;
         private Dictionary<string, XenStoreItem> items;
-        private Dictionary<string, XenStoreItemCached> cacheditems;
+        private Dictionary<string, XenStoreItemCached> cacheditems;
+
+        public void AwaitTransactionCompletion()

```

Isnt AwaitTransactionCompletion aborting the transaction, not waiting for it to be Aborted/Committed elsewhere?

---

No.  It waits to obtain the lock (with the System.Threading.Monitor.Enter() ) - which will be held by anything starting a transaction within the current session object.

Once it has the lock, it knows that the current process is not responsible for starting a transaction - so if any transaction is running, it is a zombie.  And we kill it.  

If there is no transaction running - well, we try to kill it anyway and ignore the errors (because we have no way of telling if a transaction is running)